Add role to collect operator images used in the environment

rachael-george opened the pull request:
Generate artifacts containing images used in the current deployment
Linked to task [OSP-25630](https://issues.redhat.com//browse/OSP-25630)

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

raukadah commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+      ansible.builtin.set_fact:
+        cifmw_openstack_operator_images_content: "{{ cifmw_openstack_operator_images_content | combine({ item.metadata.labels['openstack.org/operator-name'] | upper ~ '_OP_IMG': item.status.containerStatuses[1].imageID}) }}"
+
+    - name: Write images to file
+      ansible.builtin.template:
+        src: templates/operator_images.yml.j2
+        dest: "{{ cifmw_env_op_images_dir }}/artifacts/{{ cifmw_env_op_images_file }}"
```
Instead of using template, we can use ansible.builtin.copy module here like this
```suggestion
    - name: Write images to file
      ansible.builtin.copy:
        content:  |
             ----
             cifmw_install_yamls_vars:
                  {{ cifmw_install_yamls_vars_content | to_nice_yaml | replace('\n', '\n  ')}}
             cifmw_openstack_operator_images:
                   {{ cifmw_openstack_operator_images_content | to_nice_yaml | replace('\n', '\n  ')}}
              cifmw_edpm_prepare_extra_vars:
                  {{ cifmw_openstack_service_images_content | to_nice_yaml | replace('\n', '\n  ')}}
         dest: "{{ cifmw_env_op_images_dir }}/artifacts/{{ cifmw_env_op_images_file }}"
```

raukadah replied:
In order to replace `\n`, we can also use `trim` filters.

pablintino replied:
Or even better:
```yaml

- name: Write images to file
  vars:
    _content:
      cifmw_install_yamls_vars: "{{ cifmw_install_yamls_vars_content }}"
      cifmw_openstack_operator_images: "{{ cifmw_openstack_operator_images_content }}"
      cifmw_edpm_prepare_extra_vars: "{{ cifmw_openstack_service_images_content }}"
  ansible.builtin.copy:
    dest: "{{ cifmw_env_op_images_dir }}/artifacts/{{ cifmw_env_op_images_file }}"
    content: "{{ _content | to_nice_yaml }}"
```

pablintino replied:
Anyway, as I said in the review call, is really this the expected output for a role that gathers container images? A set of dicts grouped by who knows what criteria? The last thing I'd expect here is seeing variables like `cifmw_install_yamls_vars` or `cifmw_edpm_prepare_extra_vars` as keys of the dict. I know that this format is convenient as it will be easy to consume, but common, let's try to make roles decoupled from each other.

raukadah commented on ci_framework/roles/env_op_images/templates/operator_images.yml.j2:
```diff
@@ -0,0 +1,6 @@
+cifmw_install_yamls_vars:
+  {{ cifmw_install_yamls_vars_content | to_nice_yaml | replace('\n', '\n  ')}}
+cifmw_openstack_operator_images:
+  {{ cifmw_openstack_operator_images_content | to_nice_yaml | replace('\n', '\n  ')}}
+cifmw_edpm_prepare_extra_vars:
+  {{ cifmw_openstack_service_images_content | to_nice_yaml | replace('\n', '\n  ')}}
```
We can drop this line in favor of above copy comment.

raukadah reviewed (commented):
left few comments based on the suggestions from Monday review call.

pablintino commented on ci_framework/roles/env_op_images/molecule/default/converge.yml:
```diff
@@ -0,0 +1,24 @@
+    cifmw_path: "{{ ansible_user_dir }}/.crc/bin:{{ ansible_user_dir }}/.crc/bin/oc:{{ ansible_user_dir }}/bin:{{ ansible_env.PATH }}"
+    cifmw_openshift_kubeconfig: "{{ ansible_user_dir }}/.crc/machines/crc/kubeconfig"
+  roles:
+    - role: "env_op_images"
```
No verify of the generated content?

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+  environment:
+    PATH: "{{ cifmw_path | default(ansible_env.PATH) }}"
+  register: oc_installed
+  ignore_errors: true
```
If failing is not an error why not use failed_when: false? That will prevent Ansible to treat this as a failed task, that's not.

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+  environment:
+    KUBECONFIG: "{{ cifmw_openshift_kubeconfig }}"
+    PATH: "{{ cifmw_path }}"
+  ansible.builtin.shell: |
```
No need to use a shell for this. This is a single command that does the same:
` oc get OpenStackControlPlane -o=jsonpath='{.items[*].status.conditions[?(@.type=="Ready")].status}'`

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+    - name: Get images from the csv
+      ansible.builtin.shell: |
+        csv_name=$(oc get csv -n {{ cifmw_install_yamls_defaults['OPERATOR_NAMESPACE'] }} -l operators.coreos.com/openstack-operator.openstack-operators -o=jsonpath='{.items[*].metadata.name}')
+        oc get csv -n {{ cifmw_install_yamls_defaults['OPERATOR_NAMESPACE'] }} ${csv_name} -o=jsonpath='{.spec.install.spec.deployments[0].spec.template.spec.containers[1].env[*]}'
```
Selecting items in a list by index is usually a bad idea, specially when you don't manage that list insertion order. I'm against this approach. We should properly pick each item by filtering it's content.

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+        oc get csv -n {{ cifmw_install_yamls_defaults['OPERATOR_NAMESPACE'] }} ${csv_name} -o=jsonpath='{.spec.install.spec.deployments[0].spec.template.spec.containers[1].env[*]}'
+      register: csv_images
+
+    - name: Extract env variable name and images
```
This task and the next can be perfectly combined.

rachael-george replied:
Done

rachael-george commented:
@pablintino I have addressed most of the review comments. The current patch focuses on collecting the images in the environment and storing them in a yaml file in the format:
```
cifmw_csv_images:
    RELATED_IMAGE_AGENT_IMAGE_URL_DEFAULT: quay.io/openstack-k8s-operators/openstack-baremetal-operator-agent@sha256:0201cdaee9e34c700aef28e57ce778c100d5fb4c166f229c483551cb9564cc33
    RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT: quay.io/openstack-k8s-operators/openstack-ansibleee-runner@sha256:23dd6d5bfb18cace8583f0f6e4f51353ab79e9a0991170a94371039f29f44334
    ...
cifmw_openstack_operator_index_image:
    OPENSTACK_IMG: quay.io/openstack-k8s-operators/openstack-operator-index@sha256:924983377bd4d8aea1e3418ee8914f5e3ea649e002a4e5e1177fc85de86cd48f
    ...
cifmw_operator_images:
    ANSIBLEEE_OP_IMG: quay.io/openstack-k8s-operators/openstack-ansibleee-operator@sha256:d289235a35ad6e62e15a0a6b4c4d0513b316b424913a29ec9b10f987dc4bc520
    BAREMETAL_OP_IMG: quay.io/openstack-k8s-operators/openstack-baremetal-operator@sha256:2aa0124f3547081ff180f19442badd95f2b77a0cf3fb0a0799083314a70fcccb
    ...
```
Given that there is a plan to re-work the edpm_prepare and deploy roles, would it still make sense to add the functionality to deploy using the images collected here, or should I do it in a different PR after the edpm re-work is done?

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+      register: csv_images
+
+    - name: Extract env variable name and images
+      loop: "{{ csv_images.stdout| regex_findall('\\{[^}]+\\}') }}"
```
Just a style comment, is good, but for consistency, we use to put the loop fields at the very bottom of the task.

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+
+- name: Get openstack operator images
+  when:
+    - openstack_installed.stdout == "True"
```
If the previous task failed stdout may not be defined.

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+    - artifacts
+    - logs
+
+- name: Check for oc command
```
Not needed as https://github.com/openstack-k8s-operators/ci-framework/pull/772/files#diff-ea5415b13a2e225e67af327bdb8491a82cda390948f8b9f6c36eacad6ecf38edR34 will fail if oc is not present and the entire block will be skipped. I'd add a check in the block to test that `openstack_installed.rc == 0`

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+        selected_pods: "{{ pod_list.resources | rejectattr('metadata.generateName', 'contains', 'openstack-operator-index-') | rejectattr('metadata.generateName', 'contains', 'rabbitmq-cluster-operator-') }}"
+
+    - name: Add operator images to the dictionary
+      loop: "{{ selected_pods }}"
```
Just a style comment, is good, but for consistency, we use to put the loop fields at the very bottom of the task.

pablintino commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+      ansible.builtin.set_fact:
+        cifmw_openstack_operator_images_content:
+          RABBITMQ_OP_IMG: "{{ selected_pod.status.containerStatuses[0].imageID }}"
+        selected_pods: "{{ pod_list.resources | rejectattr('metadata.generateName', 'contains', 'openstack-operator-index-') | rejectattr('metadata.generateName', 'contains', 'rabbitmq-cluster-operator-') }}"
```
Just a style comment, is good, but for consistency, we try to break long jinja2 strings in multiple lines to improve readability.

chkumar246 commented on ci/config/molecule.yaml:
```diff
@@ -30,3 +30,7 @@
+- job:
+    name: cifmw-molecule-env_op_images
+    nodeset: centos-9-crc-xl
+    timeout: 5400
```
Can we drop this timeout now since we are no longer installing control plane in molecule tests?

chkumar246 commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+          OPENSTACK_IMG: "{{ selected_pod.status.containerStatuses[0].imageID }}"
+
+    - name: Get operator images and pods
+      when: not cifmw_env_op_images_dryrun
```
```suggestion
      when: not cifmw_env_op_images_dryrun | bool
```

chkumar246 commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+        openstack_installed.rc == 0 and
+        openstack_installed.stdout == "True"
+      ) or
+      cifmw_env_op_images_dryrun
```
```suggestion
      cifmw_env_op_images_dryrun | bool
```

chkumar246 commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+                        }}"
+
+    - name: Add operator images to the dictionary
+      when: not cifmw_env_op_images_dryrun
```
```suggestion
      when: not cifmw_env_op_images_dryrun | bool
```

chkumar246 commented on zuul.d/molecule.yaml:
```diff
@@ -219,6 +219,18 @@
+    name: cifmw-molecule-env_op_images
+    nodeset: centos-9-crc-xl
+    parent: cifmw-molecule-base
+    timeout: 5400
```
Timeout is still required? Can we try without this.

chkumar246 reviewed (commented):
@rachael-george Looks good, Few small comments.

raukadah commented on ci/config/molecule.yaml:
```diff
@@ -30,3 +30,7 @@
+- job:
+    name: cifmw-molecule-env_op_images
+    nodeset: centos-9-crc-xl
+    timeout: 5400
```
We can drop this one.

rachael-george replied:
Done

raukadah commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+        openstack_installed.rc == 0 and
+        openstack_installed.stdout == "True"
+      ) or
+      cifmw_env_op_images_dryrun
```
```suggestion
      cifmw_env_op_images_dryrun | bool
```

rachael-george replied:
Done

raukadah commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+                        }}"
+
+    - name: Add operator images to the dictionary
+      when: not cifmw_env_op_images_dryrun
```
```suggestion
      when: not cifmw_env_op_images_dryrun | bool
```

rachael-george replied:
Done

raukadah commented on ci_framework/roles/env_op_images/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+          OPENSTACK_IMG: "{{ selected_pod.status.containerStatuses[0].imageID }}"
+
+    - name: Get operator images and pods
+      when: not cifmw_env_op_images_dryrun
```
```suggestion
      when: not cifmw_env_op_images_dryrun | bool
```

rachael-george replied:
Done

raukadah commented on zuul.d/molecule.yaml:
```diff
@@ -219,6 +219,18 @@
+    name: cifmw-molecule-env_op_images
+    nodeset: centos-9-crc-xl
+    parent: cifmw-molecule-base
+    timeout: 5400
```
We can drop this one.

rachael-george replied:
Done
